### ISSUE TITLE
deprecate LinkContainer and move to using KnpMenu

### ISSUE THREAD

**Issue by craigh** [zikula/core#3801]
LinkContainer duplicates functionality of the MenuModule and should be changed. An extension should not use the `/Container/LinkContainer.php` structure but rather incorporate the same menus into `/Menu/AdminMenu` or something similar.

**Comment by craigh** [comment]
investigate using KnpMenu without KnpMenuBundle https://github.com/KnpLabs/KnpMenu

**Comment by craigh** [comment]
probably not - I think this would eliminate dynamic menus

**Comment by craigh** [comment]
So I think defining a menu with KNPMenu that will be used as an Admin menu is quite easily done. The question is how to define it as _the_ admin menu (as opposed to some other menu used elsewhere in the module) and then also how to know its name (ideally without any 'magic'). So my first proposal is to define it in the composer extra area. We have

// composer.json
``` json
    "extra": {
        "zikula": {
            "class": "Zikula\\BlocksModule\\ZikulaBlocksModule",
            "core-compatibility": ">=3.0.0",
            "displayname": "Blocks Module",
            "url": "blocks",
            "oldnames": [],
            "capabilities": {
                "admin": {"route": "zikulablocksmodule_admin_view"}
            },
...
```

it seems a simple matter to add something like a `menu` property

// composer.json
``` json
    "extra": {
        "zikula": {
            "class": "Zikula\\BlocksModule\\ZikulaBlocksModule",
            "core-compatibility": ">=3.0.0",
            "displayname": "Blocks Module",
            "url": "blocks",
            "oldnames": [],
            "capabilities": {
                "admin": {
                    "route": "zikulablocksmodule_admin_view",
                    "menu": {"alias": "myBlocksModuleCustomAdminMenu", "template": ~}
                }
            },
...
```

we would need to define an "api" that declares certain **arguments** that will be passed (possibly none?).
The providing module would need to define the service and tag it (auto-tagging not yet available due to requirement of defining the method & alias).

Then the AdminModule/ExtensionsModule would retrieve the menu and display it [here](https://github.com/zikula/core/blob/427d76c37afe269a533141ee4bf9740f6ea1768c/src/system/ExtensionsModule/Controller/ExtensionsInterfaceController.php#L119)

Thoughts?

**Comment by Guite** [comment]
> Thoughts?

:+1:

**Comment by Guite** [comment]
> `Zikula\Bundle \HookBundle\Listener\HooksListener` currently reacts on the `zikula.link_collector` event. This can be changed to a knp menu event (see also generator ticket referenced above).

Addition:

- remove `Zikula\ExtensionsModule\Listener\ExtensionServicesListener` and therewith the `module_dispatch.service_links` event
- remove `system/ExtensionsModule/Resources/views/Services/moduleServices.html.twig`
- instead every KnpMenu should dispatch a pre and a post event for amending it (example: [pre](https://github.com/zikula/core/blob/master/src/system/RoutesModule/Menu/Base/AbstractMenuBuilder.php#L118-L121) and [post](https://github.com/zikula/core/blob/master/src/system/RoutesModule/Menu/Base/AbstractMenuBuilder.php#L186-L189))

**Comment by Guite** [comment]
Given that there are multiple and arbitrary types possible, how about this structure:

```yaml

            "capabilities": {
                "menu_aliases": {
                    "admin": "myBlocksModuleCustomAdminMenu",
                    "user": "myBlocksModuleCustomUserMenu",
                    "account": "myBlocksModuleCustomAccountMenu",
                    "xyz": "myBlocksModuleCustomSpecialMenu"
                }
            },
```

**Comment by craigh** [comment]
no. there is no need for identification of anything but admin and user. anything else is used internally by the extension.

**Comment by Guite** [comment]
account?

**Comment by Guite** [comment]
Oh, with "user" you refer to what has been "account" earlier?

**Comment by craigh** [comment]
no - I don't think so. the "user" link is the main front-facing link for the module. link https://github.com/zikula/core/blob/master/src/system/GroupsModule/Container/LinkContainer.php#L123

and "admin" is like https://github.com/zikula/core/blob/master/src/system/GroupsModule/Container/LinkContainer.php#L82

I had forgotten about "account" (https://github.com/zikula/core/blob/master/src/system/GroupsModule/Container/LinkContainer.php#L145) so I'm not sure about that one

but any others won't be referred to by a generic name. maybe we can lose user?

**Comment by Guite** [comment]
> maybe we can lose user?

I would keep it since it is a de-facto standard.

**Comment by craigh** [comment]
hmmm. mayeb we can rething this somehow. give it a few days. I've an idea i'd like to look in to.

**Comment by craigh** [comment]
closed by #4065
